core/trino-main/src/test/java/io/trino/server/security/oauth2/TestOAuth2WebUiAuthenticationFilterWithUserinfo.java

```diff
@@ -0,0 +1,411 @@
+    private static final String TRINO_CLIENT_SECRET = "trino-secret";
+    private static final String TRINO_AUDIENCE = TRINO_CLIENT_ID;
+    private static final String ADDITIONAL_AUDIENCE = "https://external-service.com";
+    private static final String TRUSTED_CLIENT_ID = "trusted-client";
+    private static final String TRUSTED_CLIENT_SECRET = "trusted-secret";
+    private static final String UNTRUSTED_CLIENT_ID = "untrusted-client";
+    private static final String UNTRUSTED_CLIENT_SECRET = "untrusted-secret";
+    private static final String UNTRUSTED_CLIENT_AUDIENCE = "https://untrusted.com";
+
+    private final Logging logging = Logging.initialize();
+    private final OkHttpClient httpClient;
+    private TestingHydraIdentityProvider hydraIdP;
+
+    private TestingTrinoServer server;
+    private URI serverUri;
+    private URI uiUri;
+
+    public TestOAuth2WebUiAuthenticationFilterWithUserinfo()
+    {
+        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder();
+        setupInsecureSsl(httpClientBuilder);
+        httpClientBuilder.followRedirects(false);
+        httpClient = httpClientBuilder.build();
+    }
+
+    @BeforeClass
+    public void setup()
+            throws Exception
+    {
+        logging.setLevel(TestOAuth2WebUiAuthenticationFilterWithUserinfo.class.getName(), Level.DEBUG);
+
+        hydraIdP = new TestingHydraIdentityProvider(TTL_ACCESS_TOKEN_IN_SECONDS);
+        hydraIdP.start();
+        String idpUrl = "https://localhost:" + hydraIdP.getAuthPort();
+
+        server = TestingTrinoServer.builder()
+                .setCoordinator(true)
+                .setAdditionalModule(new WebUiModule())
+                .setProperties(ImmutableMap.<String, String>builder()
+                        .put("web-ui.enabled", "true")
+                        .put("web-ui.authentication.type", "oauth2")
+                        .put("http-server.https.enabled", "true")
+                        .put("http-server.https.keystore.path", Resources.getResource("cert/localhost.pem").getPath())
+                        .put("http-server.https.keystore.key", "")
+                        .put("http-server.authentication.oauth2.issuer", "https://localhost:4444/")
+                        .put("http-server.authentication.oauth2.auth-url", idpUrl + "/oauth2/auth")
+                        .put("http-server.authentication.oauth2.token-url", idpUrl + "/oauth2/token")
+                        .put("http-server.authentication.oauth2.jwks-url", idpUrl + "/.well-known/jwks.json")
+                        .put("http-server.authentication.oauth2.userinfo-url", idpUrl + "/userinfo")
+                        .put("http-server.authentication.oauth2.client-id", TRINO_CLIENT_ID)
+                        .put("http-server.authentication.oauth2.client-secret", TRINO_CLIENT_SECRET)
+                        // This is necessary as Hydra does not return `sub` from `/userinfo` for client credential grants.
```

Comment:
What does it return instead? Or more precisely how the response looks like?

Reply:
It looks like this:
```json
{
    "aud": ["trino-client"],
    "iat": 1627634164,
    "iss": "https://localhost:4444",
    "rat": -62135596800,
    "sub": ""
}
```
WIth a standard authorization code flow the subject actually has a value, but with client credentials it does not.